README.md
WiiTCPLoad
==========

A simple Qt application to send homebrew programs to a Wii running the Homebrew Channel. This is a fixed version of WiiTCPLoad from http://sourceforge.net/projects/wiitcpload/ first created by Strikeu with modifications for OS X from Tinman.

This version has been patched for qt5 by Jasjar.

Compiling for GNU/Linux:
Dependencies are the qt5 librarys required for development.
0. cd build 
1. qmake -makefile ../wiitcpload.pro
2. make
3. sudo make install


doc/CHANGELOG
WiiTCPLoad Changelog
====================

WiiTCPLoad 0.07

- Source patched to allow compiling for qt5  

WiiTCPLoad 0.06:

- Updated protocol to work with the latest HBC
- Added compression
- Added arguments


WiiTCPLoad 0.05:

- fully rewritten
- save settings in ini file


WiiTCPLoad 0.03:

- GUI fixes in main/about window. (GUI looks correct on MacOSX/Windows)
- Make install now working.
- Integration with common desktop environments. (KDE3/4, GNOME)
- Integration with konqueror, dolphin. ("Send to Homebrew/TCPLoader" action added)
- Fixed TCPLoader protocol, now application wait for accept connection.
- Improved windows version.
- Various bug fixes in source code.
- Support for FreeBSD(x86, x86_64).
- MacOSX build is now a Universal Binary.

WiiTCPLoad 0.02a:

- GUI fixes in main/about window. (GUI looks correct on Windows)
- Windows version should work now.

WiiTCPLoad 0.02:

- Support for threads.
- Various bug fixes in source code.

WiiTCPLoad 0.01:

- Support for Homebrew Channel
- Support for TCPLoader Channel


src/about.h
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <QtWidgets/QDialog>
#include "ui_about.h"

namespace Ui
{
    class AboutFormClass;
}

using namespace Ui;

class AboutForm: public QDialog {
    Q_OBJECT
public:
   AboutForm(QWidget * parent = 0);
   ~AboutForm();

private:
   Ui::AboutFormClass *ui;

public slots:
   void slotConfirmButton();

};


src/main.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

//#include <QtGui/QApplication>
#include <QtWidgets/QApplication>
#include "mainwindow.h"

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);
    MainWindow w;
    w.show();
    return a.exec();
}


src/about.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "about.h"

AboutForm::AboutForm(QWidget *parent) :QDialog(parent), ui(new Ui::AboutFormClass)
{
    ui->setupUi(this);
    connect(ui->confirmBtn, SIGNAL(clicked()), this, SLOT(slotConfirmButton()));
}

AboutForm::~AboutForm()
{
    delete ui;
}

void AboutForm::slotConfirmButton()
{
   close();
}


src/mainwindow.h
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef MAINWINDOW_H
#define MAINWINDOW_H

//#include <QtGui/QMainWindow>
#include <QtWidgets/QMainWindow>
#include "qwiistreamthread.h"
#include "about.h"

namespace Ui
{
    class MainWindowClass;
}

class MainWindow : public QMainWindow
{
    Q_OBJECT

public:
    MainWindow(QWidget *parent = 0);
    ~MainWindow();

    QWiiStreamThread *wiiStreamThread;
    AboutForm *aboutForm;

    QString appPath, configPath;

public slots:
    void transferDone();
    void transferFail(QString errorName);
    void progressBarPosition(int value);

private:
    Ui::MainWindowClass *ui;

private slots:
    void on_streamButton_clicked();

    void actionQuit(){ close(); };
    void actionAbout();
    void openFile();
};

#endif // MAINWINDOW_H


src/qwiistreamthread.h
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef QWIISTREAMTHREAD_H
#define QWIISTREAMTHREAD_H

#include <QThread>
#include <QTcpSocket>
#include <QFile>

const qint16 HOMEBREW_PORT = 4299;

class QWiiStreamThread : public QThread
{
    Q_OBJECT
public:
    QWiiStreamThread(QString host, QByteArray *data, int size, int size_compressed, QByteArray *args, int args_len);
   ~QWiiStreamThread();

private:
   QString hostname;
   QTcpSocket *tcpSocket;
   QByteArray *dataStream;
   QDataStream *readData;
   int full_size;
   int compressed_size;
   QByteArray *arguments;
   int arguments_size;
   QString errorName;

   qint8 status;
   qint64 readed, total;


private slots:
   void bytesWritten(qint64 value);

public slots:
   void slotConnected();
   void slotError(QAbstractSocket::SocketError socketError);

protected:
   void run();

signals:
   void transferDone();
   void transferFail(QString error);
   void progressBarPosition(int);

};

#endif // QWIISTREAMTHREAD_H


src/qwiistreamthread.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <QtCore/qglobal.h>




#include "qwiistreamthread.h"

#include <QAbstractSocket>
#include <QMetaType>
#include <QByteArray>
#include <QDataStream>

const qint8 ok = 0x00;
const qint8 fail = 0x01;

const qint16 wait_time = 500;

QWiiStreamThread::QWiiStreamThread(QString host, QByteArray *data, int size, int size_compressed, QByteArray *args, int args_len)
{
    qRegisterMetaType<QAbstractSocket::SocketError>("QAbstractSocket::SocketError");
    qRegisterMetaType<QAbstractSocket::SocketState>("QAbstractSocket::SocketState");
    this->setTerminationEnabled(true);
    hostname = host;
    status = ok;
    dataStream = data;
    full_size = size;
    compressed_size = size_compressed;
    arguments = args;
    arguments_size = args_len;
    errorName = QString("");
}

QWiiStreamThread::~QWiiStreamThread()
{
}

void QWiiStreamThread::run()
{
    tcpSocket = new QTcpSocket();
    connect(tcpSocket, SIGNAL(connected()), this, SLOT(slotConnected()));
    connect(tcpSocket, SIGNAL(error(QAbstractSocket::SocketError)), this, SLOT(slotError(QAbstractSocket::SocketError)));
    tcpSocket->connectToHost(hostname, HOMEBREW_PORT);
    exec();
    if (status == ok)
        tcpSocket->waitForDisconnected(3000);
    errorName = tcpSocket->errorString();
    delete tcpSocket;

    if (status == ok)
        transferDone(); else
            transferFail(errorName);
}

void QWiiStreamThread::slotConnected()
{
    readData = new QDataStream(*dataStream);
    readed = 0, total = 0;
    unsigned char datagram[4];
    char header[4];
    header[0] = 'H';
    header[1] = 'A';
    header[2] = 'X';
    header[3] = 'X';
    tcpSocket->write((const char *) &header, sizeof(header));
    tcpSocket->flush();
    datagram[0] = 0x00;
    datagram[1] = 0x05;
    datagram[2] = (arguments_size >> 0x08);
    datagram[3] = arguments_size;
    tcpSocket->write((const char *) &datagram, sizeof(datagram));
    tcpSocket->flush();
    datagram[0] = (compressed_size >> 0x18);
    datagram[1] = (compressed_size >> 0x10);
    datagram[2] = (compressed_size >> 0x08);
    datagram[3] = compressed_size;
    tcpSocket->write((const char *) &datagram, sizeof(datagram));
    tcpSocket->flush();
    datagram[0] = (full_size >> 0x18);
    datagram[1] = (full_size >> 0x10);
    datagram[2] = (full_size >> 0x08);
    datagram[3] = full_size;
    tcpSocket->write((const char *) &datagram, sizeof(datagram));
    connect(tcpSocket, SIGNAL(bytesWritten(qint64)), this, SLOT(bytesWritten(qint64)));
    tcpSocket->flush();
}

void QWiiStreamThread::bytesWritten(qint64)
{
    emit progressBarPosition(total);
    if (readData->atEnd() == true)
    {
        if(arguments_size)
            tcpSocket->write(*arguments, arguments_size);
        delete readData;
        disconnect(tcpSocket, 0, 0, 0);
        status = ok;
        quit();
        return;
    }
    char buffer[4096];
    readed = readData->readRawData(buffer, sizeof(buffer));
    total += readed;
    tcpSocket->write(buffer, readed);
}

void QWiiStreamThread::slotError(QAbstractSocket::SocketError)
{
    errorName = tcpSocket->errorString();
    status = fail;
    quit();
}


src/mainwindow.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Bartlomiej Burdukiewicz                         *
 *   dev.strikeu@gmail.com                                                 *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include "mainwindow.h"
#include "ui_mainwindow.h"

#include <QFile>
#include <QMessageBox>
#include <QFileDialog>
#include <QSettings>
#include <QByteArray>

const QString FILE_NOT_FOUND = "%1: File not found !";
const QString CANT_OPEN_TO_READ = "%1: Cannot open file to read !";
const QString FILE_IS_EMPTY = "%1: File is empty !";

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindowClass)
{
    QDir *directory = new QDir;

    appPath = QDir::homePath() + QString("/.wiitcpload/");
    configPath = appPath + QString("config.ini");

    if (directory->exists(appPath) != true)
    {
        if (directory->mkdir(appPath) != true)
        {
            QMessageBox::critical(this, tr("Critical"), QString("Can't create %1 directory").arg(appPath));
            return;
        }
    }

    delete directory;

    ui->setupUi(this);

    if (QCoreApplication::arguments().count() > 1)
        ui->fileEdit->setText(QCoreApplication::arguments().at(1));


    QSettings *configFile = new QSettings(configPath, QSettings::IniFormat);

    configFile->beginGroup("settings");
    ui->hostEdit->setText(configFile->value("Hostname", "").toString());
    if (ui->fileEdit->text().count() == 0)
        ui->fileEdit->setText(configFile->value("Filename", "").toString());
    if (ui->argEdit->text().count() == 0)
        ui->argEdit->setText(configFile->value("Arguments", "").toString());
    configFile->endGroup();

    delete configFile;

    ui->actionQuit->setShortcut(QKeySequence::Close);

    connect(ui->actionQuit, SIGNAL(triggered()), this, SLOT(actionQuit()));
    connect(ui->actionAbout, SIGNAL(triggered()), this, SLOT(actionAbout()));
    connect(ui->openFile, SIGNAL(clicked()), this, SLOT(openFile()));
}

MainWindow::~MainWindow()
{
    QSettings *configFile = new QSettings(configPath, QSettings::IniFormat);

    configFile->beginGroup("settings");
    configFile->setValue("Hostname", ui->hostEdit->text());
    configFile->setValue("Filename", ui->fileEdit->text());
    configFile->setValue("Arguments", ui->argEdit->text());
    configFile->endGroup();

    delete configFile;

    delete ui;
}

void MainWindow::actionAbout()
{
    aboutForm = new AboutForm(this);
    aboutForm->exec();
    delete aboutForm;
}

void MainWindow::openFile()
{
    QFileDialog *fileOpenDialog = new QFileDialog(this);
    QString fileName = fileOpenDialog->getOpenFileName();
    if (fileName.count() != 0)
        ui->fileEdit->setText(fileName);
    delete fileOpenDialog;

}

void MainWindow::progressBarPosition(int value)
{
    ui->progressBar->setValue(value);
}

void MainWindow::transferDone()
{
    QMessageBox::information(this, "Information", "Data written successful");
    ui->streamButton->setEnabled(true);
    ui->progressBar->setEnabled(false);
    ui->progressBar->setValue(0);
}

void MainWindow::transferFail(QString errorName)
{
    QMessageBox::critical(this, "Critical", errorName);
    ui->streamButton->setEnabled(true);
    ui->progressBar->setEnabled(false);
    ui->progressBar->setValue(0);
}

void MainWindow::on_streamButton_clicked()
{
    if (QFile::exists(ui->fileEdit->text()) != true)
    {
        QMessageBox::critical(this, "Critical", QString(FILE_NOT_FOUND).arg(ui->fileEdit->text()));
        return;
    }

    QFile *file = new QFile(ui->fileEdit->text());
    if (!file->open(QIODevice::ReadOnly))
    {
        delete file;
        QMessageBox::critical(this, "Critical", QString(CANT_OPEN_TO_READ).arg(ui->fileEdit->text()));
        return;
    }

    if (file->size() == 0)
    {
        file->close();
        delete file;
        QMessageBox::critical(this, "Critical", QString(FILE_IS_EMPTY).arg(ui->fileEdit->text()));
    }

    QByteArray *data = new QByteArray();
    *data = file->readAll();
    file->close();
    delete file;

    QByteArray *data_compressed = new QByteArray();
    *data_compressed = qCompress(*data).remove(0, 4);

    int size = data->size();
    int size_compressed = data_compressed->size();
    if(size_compressed < size) {
        delete data;
        data = data_compressed;
    }
    else {
        delete data_compressed;
        size_compressed = size;
    }

    QByteArray *args = new QByteArray();
    //*args = ui->argEdit->text().toAscii(); //toAscii() has been depreciated in qt5
    *args = ui->argEdit->text().toLatin1(); //toAscii() has been replaced by toLatin1()
    int args_len = 0;
    if(!args->isEmpty())
        args_len = args->size() + 1;

    ui->progressBar->setMaximum(size_compressed);
    ui->progressBar->setEnabled(true);

    wiiStreamThread = new QWiiStreamThread(ui->hostEdit->text(), data, size, size_compressed, args, args_len);
    connect(wiiStreamThread, SIGNAL(transferFail(QString)), this, SLOT(transferFail(QString)));
    connect(wiiStreamThread, SIGNAL(transferDone()), this, SLOT(transferDone()));
    connect(wiiStreamThread, SIGNAL(progressBarPosition(int)), this, SLOT(progressBarPosition(int)));
    wiiStreamThread->start();
    ui->streamButton->setEnabled(false);
}
